test/test_epoll.cpp

```diff
@@ -351,3 +381,191 @@ TEST(CEPoll, NotifyConnectionBreak)
+        EXPECT_EQ(epoll.remove_usock(epoll_id, client_sock), 0);
+    }
+    catch (CUDTException &ex)
+    {
+        cerr << ex.getErrorMessage() << endl;
+        throw;
+    }
+
+    try
+    {
+        EXPECT_EQ(epoll.release(epoll_id), 0);
+    }
+    catch (CUDTException &ex)
+    {
+        cerr << ex.getErrorMessage() << endl;
+        throw;
+    }
+
+    EXPECT_EQ(srt_cleanup(), 0);
+}
+
+
+TEST(CEPoll, HandleEpollNoEvent)
+{
+    ASSERT_EQ(srt_startup(), 0);
+
+    SRTSOCKET client_sock = srt_socket(AF_INET, SOCK_DGRAM, 0);
+    EXPECT_NE(client_sock, SRT_ERROR);
+
+    const int yes = 1;
+    const int no  = 0;
+    EXPECT_NE(srt_setsockopt (client_sock, 0, SRTO_RCVSYN,    &no,  sizeof no),  SRT_ERROR); // for async connect
+    EXPECT_NE(srt_setsockopt (client_sock, 0, SRTO_SNDSYN,    &no,  sizeof no),  SRT_ERROR); // for async connect
+    EXPECT_NE(srt_setsockflag(client_sock,    SRTO_SENDER,    &yes, sizeof yes), SRT_ERROR);
+    EXPECT_NE(srt_setsockopt (client_sock, 0, SRTO_TSBPDMODE, &yes, sizeof yes), SRT_ERROR);
+
+    CEPoll epoll;
+    const int epoll_id = epoll.create();
+    ASSERT_GE(epoll_id, 0);
+
+    const int epoll_out = SRT_EPOLL_OUT | SRT_EPOLL_ERR;
+    ASSERT_NE(epoll.add_usock(epoll_id, client_sock, &epoll_out), SRT_ERROR);
+
+    /*
+       Subscribe to the event, but do not report them.
+       In result you should have no events reported.
+    set<int> epoll_ids = { epoll_id };
+
+    epoll.update_events(client_sock, epoll_ids, SRT_EPOLL_ERR, true);
+    */
+
+    SRT_EPOLL_EVENT fds[1024];
+
+    // Use timeout 0 because with -1 this call would hang up
+    int result = epoll.uwait(epoll_id, fds, 1024, 0);
+    ASSERT_EQ(result, 0); 
+
+    try
+    {
+        EXPECT_EQ(epoll.remove_usock(epoll_id, client_sock), 0);
+    }
+    catch (CUDTException &ex)
+    {
+        cerr << ex.getErrorMessage() << endl;
+        throw;
+    }
+
+    try
+    {
+        EXPECT_EQ(epoll.release(epoll_id), 0);
+    }
+    catch (CUDTException &ex)
+    {
+        cerr << ex.getErrorMessage() << endl;
+        throw;
+    }
+
+    EXPECT_EQ(srt_cleanup(), 0);
+}
+
+TEST(CEPoll, ThreadedUpdate)
+{
+    ASSERT_EQ(srt_startup(), 0);
+
+    SRTSOCKET client_sock = srt_socket(AF_INET, SOCK_DGRAM, 0);
+    EXPECT_NE(client_sock, SRT_ERROR);
+
+    const int yes = 1;
+    const int no  = 0;
+    EXPECT_NE(srt_setsockopt (client_sock, 0, SRTO_RCVSYN,    &no,  sizeof no),  SRT_ERROR); // for async connect
+    EXPECT_NE(srt_setsockopt (client_sock, 0, SRTO_SNDSYN,    &no,  sizeof no),  SRT_ERROR); // for async connect
+
+    CEPoll epoll;
+    const int epoll_id = epoll.create();
+    ASSERT_GE(epoll_id, 0);
+    ASSERT_EQ(epoll.setflags(epoll_id, SRT_EPOLL_ENABLE_EMPTY), 0);
+
+    thread td = thread( [&epoll, epoll_id, client_sock]()
+    {
+        cerr << "Spawned thread to add sockets to eid (wait 1s to order execution)\n";
+        this_thread::sleep_for(chrono::seconds(1)); // Make sure that uwait will be called as first
```

Comment:
I had the same idea, but - guess what - it's impossible. There's no way to trigger the signal. Simply because what is required in this test is that things happen in the following order:

* You call the `uwait` function. This function should see that the EID is empty and react appropriately (that is, not react at all in this case). This check is actually done in the loop, the test must make sure that this has happened AT LEAST ONCE.
* Enter the waiting loop. Next checks for the above may occur, it doesn't matter
* THREAD SWITCH
* The second thread suddenly subscribes the EID to the `client_sock` socket
* ... then it issues an event on this socket
* and can exit
* THREAD SWITCH
* in the next iteration of the loop the `uwait` function realizes that it now has some subscribed sockets - surprise! - and there was an event reported on it

In order to use signal, you'd have to somehow magically recognize that the `uwait` function has run the internal loop at least once and has done the check. I can't think of any, except adding some test-only intrusive code that would raise the signal.